File: tests/corpus/concept.lua
-- vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
-- Re-sync when vidya adds a Lua reference sample.
--
-- Lexing + parsing of arithmetic expressions, idiomatic Lua.

--[[
Multi-line block comment using Lua's --[[ ... ]] long-bracket
form. The simple version (no = padding) is recognised by this
grammar; the variable-padded form (--[==[ ... ]==]) is a
documented gap.
]]

-- ── Tokens ───────────────────────────────────────────────────────────

local Token = {}
Token.NUMBER = "number"
Token.PLUS   = "plus"
Token.MINUS  = "minus"
Token.STAR   = "star"
Token.SLASH  = "slash"
Token.LPAREN = "lparen"
Token.RPAREN = "rparen"
Token.EOF    = "eof"

local function make_token(kind, text, pos)
    return { kind = kind, text = text, pos = pos }
end

-- ── Lexer ────────────────────────────────────────────────────────────

local Lexer = {}
Lexer.__index = Lexer

function Lexer.new(source)
    local self = setmetatable({}, Lexer)
    self.source = source
    self.pos = 1
    return self
end

local function is_digit(c)
    return c and c >= '0' and c <= '9'
end

function Lexer:peek()
    return self.source:sub(self.pos, self.pos)
end

function Lexer:skip_whitespace()
    while self.pos <= #self.source and self:peek():match("%s") do
        self.pos = self.pos + 1
    end
end

function Lexer:next()
    self:skip_whitespace()
    local start = self.pos

    if self.pos > #self.source then
        return make_token(Token.EOF, "", self.pos)
    end

    local c = self:peek()
    if c == "+" then self.pos = self.pos + 1; return make_token(Token.PLUS,   "+", start) end
    if c == "-" then self.pos = self.pos + 1; return make_token(Token.MINUS,  "-", start) end
    if c == "*" then self.pos = self.pos + 1; return make_token(Token.STAR,   "*", start) end
    if c == "/" then self.pos = self.pos + 1; return make_token(Token.SLASH,  "/", start) end
    if c == "(" then self.pos = self.pos + 1; return make_token(Token.LPAREN, "(", start) end
    if c == ")" then self.pos = self.pos + 1; return make_token(Token.RPAREN, ")", start) end

    if is_digit(c) then
        while self.pos <= #self.source and is_digit(self:peek()) do
            self.pos = self.pos + 1
        end
        return make_token(Token.NUMBER, self.source:sub(start, self.pos - 1), start)
    end

    error(string.format("unexpected character %q at %d", c, self.pos))
end

-- ── AST ──────────────────────────────────────────────────────────────

local function number_expr(value)
    return { kind = "number", value = value }
end

local function unary_expr(op, operand)
    return { kind = "unary", op = op, operand = operand }
end

local function binop_expr(op, left, right)
    return { kind = "binop", op = op, left = left, right = right }
end

local function eval(e)
    if e.kind == "number" then return e.value end
    if e.kind == "unary"  then return -eval(e.operand) end
    if e.kind == "binop"  then
        local a = eval(e.left)
        local b = eval(e.right)
        if e.op == "+" then return a + b end
        if e.op == "-" then return a - b end
        if e.op == "*" then return a * b end
        if e.op == "/" then return a // b end
        error("bad op " .. e.op)
    end
    error("bad expr kind " .. tostring(e.kind))
end

-- ── Pratt parser ─────────────────────────────────────────────────────

local INFIX_BP = {
    [Token.PLUS]  = 1,
    [Token.MINUS] = 1,
    [Token.STAR]  = 3,
    [Token.SLASH] = 3,
}

local Parser = {}
Parser.__index = Parser

function Parser.new(source)
    local self = setmetatable({}, Parser)
    self.lexer = Lexer.new(source)
    self.current = self.lexer:next()
    return self
end

function Parser:advance()
    self.current = self.lexer:next()
end

function Parser:parse_expr(min_bp)
    local left = self:parse_primary()
    while true do
        local bp = INFIX_BP[self.current.kind] or 0
        if bp == 0 or bp < min_bp then break end
        local op = self.current.text
        self:advance()
        local right = self:parse_expr(bp + 1)
        left = binop_expr(op, left, right)
    end
    return left
end

function Parser:parse_primary()
    if self.current.kind == Token.MINUS then
        self:advance()
        return unary_expr("-", self:parse_primary())
    end
    if self.current.kind == Token.LPAREN then
        self:advance()
        local inner = self:parse_expr(0)
        if self.current.kind ~= Token.RPAREN then
            error("expected )")
        end
        self:advance()
        return inner
    end
    if self.current.kind == Token.NUMBER then
        local v = tonumber(self.current.text)
        self:advance()
        return number_expr(v)
    end
    error("unexpected token " .. self.current.kind)
end

-- ── Demo ─────────────────────────────────────────────────────────────

local cases = {
    { "1 + 2",       3 },
    { "2 * 3 + 4",   10 },
    { "(1 + 2) * 3", 9 },
    { "-5 + 3",      -2 },
}

for _, case in ipairs(cases) do
    local input, expected = case[1], case[2]
    local got = eval(Parser.new(input):parse_expr(0))
    print(string.format("%-15s => %d (expected %d)", input, got, expected))
    assert(got == expected, "mismatch at " .. input)
end

File: tests/corpus/concept.swift
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a Swift reference sample.
//
// Lexing + parsing of arithmetic expressions, idiomatic Swift.

import Foundation

/*
 * Multi-line block comment using Swift's slash-star pair form.
 * Swift block comments are nestable per the spec; vyakarana's
 * pair rule is greedy and closes at the first slash-star
 * close. Documented gap, same shape as Rust.
 */

// ── Tokens ───────────────────────────────────────────────────────────

enum TokenKind {
    case number
    case plus
    case minus
    case star
    case slash
    case lparen
    case rparen
    case eof
}

struct Token {
    let kind: TokenKind
    let text: String
    let pos: Int
}

// ── Lexer ────────────────────────────────────────────────────────────

final class Lexer {
    private let source: [Character]
    private var pos: Int = 0

    init(_ source: String) {
        self.source = Array(source)
    }

    private func peek() -> Character? {
        guard pos < source.count else { return nil }
        return source[pos]
    }

    private func skipWhitespace() {
        while let c = peek(), c.isWhitespace {
            pos += 1
        }
    }

    func next() -> Token {
        skipWhitespace()
        let start = pos

        guard let c = peek() else {
            return Token(kind: .eof, text: "", pos: pos)
        }

        switch c {
        case "+": pos += 1; return Token(kind: .plus,   text: "+", pos: start)
        case "-": pos += 1; return Token(kind: .minus,  text: "-", pos: start)
        case "*": pos += 1; return Token(kind: .star,   text: "*", pos: start)
        case "/": pos += 1; return Token(kind: .slash,  text: "/", pos: start)
        case "(": pos += 1; return Token(kind: .lparen, text: "(", pos: start)
        case ")": pos += 1; return Token(kind: .rparen, text: ")", pos: start)
        default: break
        }

        if c.isNumber {
            while let d = peek(), d.isNumber {
                pos += 1
            }
            let text = String(source[start..<pos])
            return Token(kind: .number, text: text, pos: start)
        }

        fatalError("unexpected character '\(c)' at \(pos)")
    }
}

// ── AST ──────────────────────────────────────────────────────────────

indirect enum Expr {
    case number(Int64)
    case unary(Character, Expr)
    case binop(Character, Expr, Expr)

    func eval() -> Int64 {
        switch self {
        case .number(let v): return v
        case .unary(_, let e): return -e.eval()
        case .binop(let op, let l, let r):
            let a = l.eval()
            let b = r.eval()
            switch op {
            case "+": return a + b
            case "-": return a - b
            case "*": return a * b
            case "/": return a / b
            default:  fatalError("bad op \(op)")
            }
        }
    }
}

// ── Pratt parser ─────────────────────────────────────────────────────

final class Parser {
    private let lexer: Lexer
    private var current: Token

    init(_ source: String) {
        let l = Lexer(source)
        self.lexer = l
        self.current = l.next()
    }

    private func advance() {
        current = lexer.next()
    }

    private static func infixBp(_ k: TokenKind) -> Int {
        switch k {
        case .plus, .minus:  return 1
        case .star, .slash:  return 3
        default:             return 0
        }
    }

    func parseExpr(_ minBp: Int) -> Expr {
        var left = parsePrimary()
        while true {
            let bp = Self.infixBp(current.kind)
            if bp == 0 || bp < minBp { break }
            let op = Character(current.text)
            advance()
            let right = parseExpr(bp + 1)
            left = .binop(op, left, right)
        }
        return left
    }

    private func parsePrimary() -> Expr {
        if current.kind == .minus {
            advance()
            return .unary("-", parsePrimary())
        }
        if current.kind == .lparen {
            advance()
            let inner = parseExpr(0)
            guard current.kind == .rparen else {
                fatalError("expected )")
            }
            advance()
            return inner
        }
        if current.kind == .number {
            let v = Int64(current.text) ?? 0
            advance()
            return .number(v)
        }
        fatalError("unexpected token")
    }
}

// ── Demo ─────────────────────────────────────────────────────────────

let cases: [(String, Int64)] = [
    ("1 + 2",       3),
    ("2 * 3 + 4",   10),
    ("(1 + 2) * 3", 9),
    ("-5 + 3",      -2),
]

for (input, expected) in cases {
    let got = Parser(input).parseExpr(0).eval()
    let line = """
        \(input.padding(toLength: 15, withPad: " ", startingAt: 0)) \
        => \(got) (expected \(expected))
        """
    print(line)
    precondition(got == expected, "mismatch at \"\(input)\"")
}
